Implement Equals method for some types

oceanstuck opened the pull request:
A couple applications I'm working on require the ability to check value equality for dialogs and procedures and the types of their members by extension. Idk of any applications that rely on the default class behavior of checking reference equality so I'm just assuming it's fine to do this.

oceanstuck committed:
implement value equality for some types

oceanstuck committed:
Merge branch 'master' of https://github.com/oceanstuck/Atlus-Script-Tools

oceanstuck committed:
value equality fixes and tests

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/IDialog.cs:
```diff
@@ -1,24 +1,101 @@
 /// Common interface for message script dialog windows.
 /// </summary>
-public interface IDialog : IEnumerable<TokenText>
+public abstract class IDialog : IEnumerable<TokenText>, IEquatable<IDialog>
```
You can't just change an interface to an abstract class without at least changing the name. Additionally, is it really necessary to change it from an interface to abstract class in the first place?

oceanstuck replied:
I did try to keep this and ISpeaker as interfaces but ran into trouble trying to implement a static `Equals(IDialog, IDialog)` member (trying to include a default implementation in the interface itself throws CS8701 ("Target runtime doesn't support default interface implementation")). Looking at it with fresh eyes though, I think I overcomplicated the problem? Implementing a static `Equals(MessageDialog, IDialog` and `Equals(SelectionDialog, IDialog` in the respective derived classes seems to work just fine, so I'll just revert to interface unless some other issue comes up.

oceanstuck replied:
Ok it's coming back to me, and also the derived static equals trick doesn't actually work. Operator overloads require one of the args to be of the type implementing the overload, which combined with the above means I can't do eg. `operator!=(ISpeaker, ISpeaker)` and therefore `this.Speaker != other.Speaker` doesn't work. I can get around that specific example while comparing dialogs by just doing `this.Speaker.Equals(other.Speaker)` but not being able to use == and != feels just...awful. Looking things up online, it seems like default interface implementation requires C# 8 or later but in the project file LangVersion is set to latest :|

oceanstuck replied:
Ah I see, .NET Standard 2.0 defaults to C# 7.3. It might have to be an abstract class after all if I want operator overloads, though out of curiosity what AST-using applications are still on .NET Framework anyway?

oceanstuck replied:
Abstract classes are renamed

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/IDialog.cs:
```diff
@@ -1,24 +1,101 @@
+
+    public override int GetHashCode()
+    {
+        int hashCode = -1358287131;
```
I believe HashCode methods are to be used for this in modern .NET

oceanstuck replied:
I just let Visual Studio generate the GetHashCode methods tbh, I wasn't sure what the best practice was

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/IDialog.cs:
```diff
@@ -1,24 +1,101 @@
+    public static bool operator!=(IDialog x, IDialog y) => !Equals(x, y);
+}
+
+/*public class DialogComparer : EqualityComparer<IDialog>
```
Need to clean this up.

oceanstuck replied:
comments removed

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/ISpeaker.cs:
```diff
@@ -1,12 +1,48 @@
 /// Common interface for dialogue message speakers.
 /// </summary>
-public interface ISpeaker
+public abstract class ISpeaker : IEquatable<ISpeaker>
```
Same as previous interface comment.

oceanstuck replied:
same deal as the other interface

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/ISpeaker.cs:
```diff
@@ -1,12 +1,48 @@
+        switch (Kind)
+        {
+            case SpeakerKind.Named:
+                return ((NamedSpeaker)this).Equals(other as NamedSpeaker);
```
Upcasting from base is ugly here, this could be handled better by providing/overriding Equals on the subclasses

oceanstuck replied:
done

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/IToken.cs:
```diff
@@ -10,3 +13,43 @@ public interface IToken
+            case TokenKind.NewLine:
+                return true;
+            case TokenKind.Function:
+                return ((FunctionToken)x).Equals((FunctionToken)y);
```
Same as previous upcast comment

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/IToken.cs:
```diff
@@ -10,3 +13,43 @@ public interface IToken
+        switch (obj.Kind)
+        {
+            case TokenKind.NewLine:
+                return ((NewLineToken)obj).GetHashCode();
```
Same as previous upcast comment

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/MessageDialog.cs:
```diff
@@ -8,12 +8,12 @@ namespace AtlusScriptLibrary.MessageScriptLanguage;
     /// Gets the text identifier of this dialog window.
     /// </summary>
-    public string Name { get; set; }
+    //public string Name { get; set; }
```
Why?

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/MessageDialog.cs:
```diff
@@ -112,15 +112,66 @@ public override string ToString()
     }
 
-    IEnumerator IEnumerable.GetEnumerator()
+    /*IEnumerator IEnumerable.GetEnumerator()
```
Needs to be cleaned up

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/MessageDialog.cs:
```diff
@@ -112,15 +112,66 @@ public override string ToString()
+        return (Pages.SequenceEqual(obj.Pages));
+    }
+
+    /*public bool Equals(IDialog obj)
```
Needs to be cleaned up

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/MessageDialog.cs:
```diff
@@ -112,15 +112,66 @@ public override string ToString()
+
+    public override int GetHashCode()
+    {
+        int hashCode = 1327504663;
```
Same as previous HashCode comment

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/MessageDialog.cs:
```diff
@@ -112,15 +112,66 @@ public override string ToString()
+    public static bool operator ==(MessageDialog x, MessageDialog y) => Equals(x, y);
+    public static bool operator !=(MessageDialog x, MessageDialog y) => !Equals(x, y);
+
+    /*public static bool operator ==(MessageDialog x, IDialog y) => x.Equals(y);
```
Needs to be cleaned up

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/SelectionDialog.cs:
```diff
@@ -7,12 +8,12 @@ namespace AtlusScriptLibrary.MessageScriptLanguage;
     /// Gets the text identifier of this dialog.
     /// </summary>
-    public string Name { get; set; }
+    //public string Name { get; set; }
```
Why?

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/SelectionDialog.cs:
```diff
@@ -64,15 +65,66 @@ public SelectionDialog(string identifier, SelectionDialogPattern pattern, params
     }
 
-    IEnumerator IEnumerable.GetEnumerator()
+    /*IEnumerator IEnumerable.GetEnumerator()
```
Needs to be cleaned up

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/SelectionDialog.cs:
```diff
@@ -64,15 +65,66 @@ public SelectionDialog(string identifier, SelectionDialogPattern pattern, params
+        return (Options.SequenceEqual(obj.Options));
+    }
+
+    /*public bool Equals(IDialog obj)
```
Needs to be cleaned up

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/SelectionDialog.cs:
```diff
@@ -64,15 +65,66 @@ public SelectionDialog(string identifier, SelectionDialogPattern pattern, params
+
+    public override int GetHashCode()
+    {
+        int hashCode = -95244068;
```
HashCode.Combine

oceanstuck replied:
I did attempt this but doing so throws CS0234 ("The type or namespace name 'HashCode' does not exist in the namespace 'System' (are you missing an assembly reference?)"). Intellisense gives me a warning that the HashCode struct is available in .NET 8 but not .NET Standard 2.0, idk if there's a particular reason AST uses the latter so I'm gonna leave it for now.

tge-was-taken replied:
Need to install https://www.nuget.org/packages/Microsoft.Bcl.HashCode/ apparently. 2.0 is used for compatibility with legacy .NET framework

oceanstuck replied:
done

tge-was-taken commented on Source/AtlusScriptLibrary/MessageScriptLanguage/VariableSpeaker.cs:
```diff
@@ -25,8 +27,18 @@ public override string ToString()
+
+    public override int GetHashCode()
+    {
+        int hashCode = -1061823130;
```
HashCode combine

oceanstuck replied:
same as the other hash code comment

tge-was-taken commented on Source/AtlusScriptLibraryTests/FlowScriptLanguage/FlowScriptTests.cs:
```diff
@@ -247,5 +249,24 @@ public void ToBinary_ContentsShouldMatchThatOfSourceBinary_Version3BigEndian()
+        {
+            var a = FlowScript.FromFile("TestResources/Version3BigEndian.bf");
+            var b = FlowScript.FromFile("TestResources/Version3BigEndian.bf");
+            //var dialogComparer = new DialogComparer();
```
Cleanup needed

oceanstuck committed:
cleanup hash code methods and base class equality methods, rename abstract classes, remove comments

oceanstuck committed:
remove comment from FlowScriptTests

oceanstuck commented:
Don't feel like responding directly to all convos since a lot of my responses would be repetitive, but I renamed the abstract classes, cleaned up the HashCode methods and equality methods, and yeeted the commented-out code. Better?
